File: methods/Wikipedia.js
import { Promise } from 'bluebird'
import _ from 'lodash'
import { Observable } from 'rxjs/Rx'
import { mergeMap, map } from 'rxjs/operators'
import axios from 'axios'
import fs from 'fs'
let WordPOS = require('wordpos')

export default (question, answers) => {
  return Observable.from(getNouns(question).then((nouns) => {
      return Promise.map(answers, (answer) => {
        return getBody(nouns, answer)
      })
      .then((res) => transfromResults(res))
    }))
}
  // return getNouns(question).then((nouns) => {
  //   return Observable.forkJoin(
  //       ..._.map(answers, (answer) => {
  //         return Observable.from(getBody(nouns, answer))
  //       })
  //     ).map(results => {
  //         return transfromResults(results)
  //     })
  //   })
// }

let getNouns = async(question) => {
  let wordpos = new WordPOS()
  return await wordpos.getNouns(question, nouns => nouns)
}

function occurrences(string, subString, allowOverlapping) {
  string += "";
  subString += "";
  if (subString.length <= 0) return (string.length + 1);

  var n = 0,
    pos = 0,
    step = allowOverlapping ? 1 : subString.length;

  while (true) {
    pos = string.indexOf(subString, pos);
    if (pos >= 0) {
      ++n;
      pos += step;
    } else break;
  }
  return n;
}

let getBody = async(questionNouns, answer) => {
  let transAnswer = _.escape(answer)
  let transAnswer1 = answer.replace(/ /g,'_')
  let transAnswer2 = _.map(answer.split(' '), (word) => _.startCase(_.toLower(word))).join(' ')
  let url = `https://en.wikipedia.org/w/api.php?action=query&prop=extracts&format=json&titles=${transAnswer}`

  let result = await axios.get(url)

  if(Object.keys(result.data.query.pages)[0] == -1){
    transAnswer = _.startCase(_.toLower(transAnswer))
    url = `https://en.wikipedia.org/w/api.php?action=query&prop=extracts&format=json&titles=${transAnswer}`
    result = await axios.get(url)
  }
  if(Object.keys(result.data.query.pages)[0] == -1){
    transAnswer = transAnswer1
    url = `https://en.wikipedia.org/w/api.php?action=query&prop=extracts&format=json&titles=${transAnswer}`
    result = await axios.get(url)
  }
  if(Object.keys(result.data.query.pages)[0] == -1){
    transAnswer = transAnswer2
    url = `https://en.wikipedia.org/w/api.php?action=query&prop=extracts&format=json&titles=${transAnswer}`
    result = await axios.get(url)
  }

  const pageData = result.data.query.pages[Object.keys(result.data.query.pages)[0]]
  let body
  if(pageData.hasOwnProperty('missing')){
    body = ""
  } else {
    body = pageData.extract
  }

  let total = 0
  questionNouns.forEach((noun) => {
    total += occurrences(body, noun, false)
  })

  return { answer: answer, count: total }
}

let transfromResults = (results) => {
  results.sort((a,b) => {
    a = parseInt(a.count)
    b = parseInt(b.count)
    if (a < b)
      return -1;
    if (a > b)
      return 1;
    return 0;
  })

  let sum = _.reduce(results, (val, ele) => (val + parseInt(ele.count)), 0)

  let final = {
    smallest: { answer: results[0].answer, weight: 0.01 },
    middle: { answer: results[1].answer, weight: 0.09 },
    largest: { answer: results[results.length-1].answer, weight: 0.9 },
    method: 'wikipedia',
  }
  return final
}

File: methods/wikipedia.go
package methods

import (
	"io/ioutil"
	"net/http"
	"strings"
)

type PageResult struct {
	answer string
	hits   int
}

func GetWikipedia(answer string) (string, error) {
	resp, err := http.Get("http://en.wikipedia.org/w/api.php?action=query&prop=extracts&format=json&titles=" + answer)
	if err != nil {
		// handle error
		return "", err
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	response := string(body)
	return response, nil
}

func GetNumberOfOccurences(c chan<- PageResult, answer string, nouns []string) {
	count := 0
	wikiBody, _ := GetWikipedia(answer) // TODO: check for error

	if len(wikiBody) < 113*len(answer)*3 { // if its too short, try titlecase
		wikiBody, _ = GetWikipedia(strings.Title(answer))
	}

	for _, noun := range nouns {
		count += strings.Count(wikiBody, noun)
	}

	pr := PageResult{
		answer: answer,
		hits:   count,
	}

	c <- pr
}

func CalculateWeights(results [3]int) []float64 {
	total := 0
	for _, num := range results {
		total += num
	}

	var weights []float64
	for _, val := range results {
		weights = append(weights, float64(val)/float64(total))
	}

	return weights
}

// the function that grabs
var startWikipedia StartFunc = func(qna *QandA, doneChannel chan *MethodResults) {
	var numbers [3]int
	c := make(chan PageResult)

	// TODO: get an actual list of nouns, not just question minus short words
	nouns := strings.Split(qna.Question, " ")
	blen := len(nouns)
	for n := 0; n < blen; n++ {
		if len(nouns[n]) > 3 {
			nouns = append(nouns, nouns[n])
		}
	}
	nouns = nouns[blen:]

	// spin up goroutines to count up
	for i := 0; i < len(qna.Answers); i++ {
		go GetNumberOfOccurences(c, qna.Answers[i], nouns)
	}

	// consume PageResults from channel
	for r := 0; r < 3; r++ {
		result := <-c
		// find the answers array
		for i, val := range qna.Answers {
			if val == result.answer {
				numbers[i] = result.hits
			}
		}
	}

	// calculate the weights
	weights := CalculateWeights(numbers)

	methodResults := &MethodResults{
		Name:    "Wikipedia",
		Results: weights,
	}

	doneChannel <- methodResults
}
